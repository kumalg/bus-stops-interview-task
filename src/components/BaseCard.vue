<template>
  <div class="card" :class="{ dotted, 'no-paddings': withoutPaddings }">
    <div v-if="title" class="card-header">
      <h2>{{ title }}</h2>
    </div>
    <div class="card-body">
      <slot />
    </div>
  </div>
</template>

<script setup lang="ts">
withDefaults(defineProps<{
  dotted?: boolean;
  withoutPaddings?: boolean,
  title?: string
}>(), {
  dotted: false,
  withoutPaddings: false,
  title: ''
})
</script>

<style lang="scss" scoped>
.card {
  background-color: $color-gray-000;
  border-radius: 0.25rem;
  display: flex;
  flex-direction: column;

  &.dotted {
    border: 2px solid red;
  }

  .card-header {
    padding: $card-padding $card-padding 0.5rem $card-padding;
  }

  &:not(.no-paddings) {
    .card-body {
      padding: $card-padding;
      flex: 1;
    }
  }
}
</style>
