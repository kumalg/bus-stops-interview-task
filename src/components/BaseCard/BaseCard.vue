<template>
    <div class="card" :class="{ 'no-paddings': withoutPaddings }">
        <div v-if="title" class="card-header">
            <h2>{{ title }}</h2>
        </div>
        <div class="card-body">
            <slot />
        </div>
    </div>
</template>

<script setup lang="ts">
withDefaults(
    defineProps<{
        withoutPaddings?: boolean;
        title?: string;
    }>(),
    {
        withoutPaddings: false,
        title: ''
    }
);
</script>

<style lang="scss" scoped>
.card {
    background-color: $color-gray-000;
    border-radius: 0.25rem;
    display: flex;
    flex-direction: column;
    overflow: hidden;

    .card-header {
        padding: $card-padding $card-padding 0.5rem $card-padding;
    }

    .card-body {
        min-height: 0;
        display: flex;
        flex-direction: column;
        flex: 1;
    }

    &:not(.no-paddings) {
        .card-body {
            padding: $card-padding;
        }
    }
}
</style>
