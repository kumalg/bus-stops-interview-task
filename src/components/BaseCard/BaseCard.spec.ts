import { flushPromises, shallowMount } from '@vue/test-utils';

import BaseCard from './BaseCard.vue';

const EXAMPLE_CONTENT = 'EXAMPLE_CONTENT';

describe('BaseCard', () => {
    const BaseCardFactory = (
        props: InstanceType<typeof BaseCard>['$props'] = {},
        slotContent: string = EXAMPLE_CONTENT
    ) => {
        return shallowMount(BaseCard, {
            props,
            slots: {
                default: slotContent
            }
        });
    };

    it('renders proper html markup', () => {
        const wrapper = BaseCardFactory({ title: 'TITLE' });

        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders slot conent inside card-body div', () => {
        const wrapper = BaseCardFactory();

        expect(wrapper.find('.card-body').text().trim()).toBe(EXAMPLE_CONTENT);
    });

    describe('header', () => {
        it('should not be rendered when title is not provided', () => {
            const wrapper = BaseCardFactory();

            expect(wrapper.find('.card-header').exists()).toBe(false);
            expect(wrapper.find('h2').exists()).toBe(false);
        });

        it('should be rendered when title is provided', () => {
            const wrapper = BaseCardFactory({ title: 'TITLE' });

            expect(wrapper.find('.card-header').exists()).toBe(true);
            expect(wrapper.find('h2').exists()).toBe(true);
        });
    });

    describe('no-paddings class', () => {
        it('should exist when withoutPaddings prop is true', () => {
            const wrapper = BaseCardFactory({ withoutPaddings: true });

            expect(wrapper.classes().includes('no-paddings')).toBe(true);
        });

        it('should not exist when withoutPaddings prop is false', async () => {
            const wrapper = BaseCardFactory();

            expect(wrapper.classes().includes('no-paddings')).toBe(false);
        });
    });
});
