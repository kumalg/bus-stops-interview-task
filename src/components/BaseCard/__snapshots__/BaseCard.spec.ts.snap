// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseCard renders proper html markup 1`] = `
<div
  class="card"
>
  <div
    class="card-header"
  >
    <h2>
      TITLE
    </h2>
  </div>
  <div
    class="card-body"
  >
    
    EXAMPLE_CONTENT
    
  </div>
</div>
`;
